flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/ClusterEntrypointTest.java

```diff
@@ -18,54 +18,418 @@
+                new TestingHighAvailabilityServicesBuilder()
+                        .setCloseAndCleanupAllDataFuture(closeAndCleanupAllDataFuture)
+                        .build();
+        final TestingResourceManagerFactory testingResourceManagerFactory =
+                new TestingResourceManagerFactory.Builder()
+                        .setInternalDeregisterApplicationConsumer(
+                                (ignored1, ignored2) -> deregisterFuture.complete(null))
+                        .build();
+        final TestingDispatcherRunnerFactory testingDispatcherRunnerFactory =
+                new TestingDispatcherRunnerFactory.Builder()
+                        .setShutDownFuture(dispatcherShutDownFuture)
+                        .build();
+
+        final TestingEntryPoint testingEntryPoint =
+                new TestingEntryPoint.Builder()
+                        .setConfiguration(flinkConfig)
+                        .setResourceManagerFactory(testingResourceManagerFactory)
+                        .setDispatcherRunnerFactory(testingDispatcherRunnerFactory)
+                        .setHighAvailabilityServices(testingHaService)
+                        .build();
+
+        final CompletableFuture<ApplicationStatus> appStatusFuture =
+                startClusterEntrypoint(testingEntryPoint);
+
+        dispatcherShutDownFuture.complete(ApplicationStatus.SUCCEEDED);
 
-        protected MockEntryPoint(Configuration configuration) {
+        assertThat(
+                appStatusFuture.get(TIMEOUT_MS, TimeUnit.MILLISECONDS),
+                is(ApplicationStatus.SUCCEEDED));
+        assertThat(deregisterFuture.isDone(), is(true));
+        assertThat(closeAndCleanupAllDataFuture.isDone(), is(true));
+    }
+
+    @Test
+    public void testCloseAsyncShouldBeExecutedInShutdownHook() throws Exception {
+        // This test only works on Linux and Mac OS because we are sending SIGTERM to a
+        // JAVA process via "kill {pid}"
+        assumeTrue(OperatingSystem.isLinux() || OperatingSystem.isMac());
+        final File markerFile = new File(TEMPORARY_FOLDER.getRoot(), UUID.randomUUID() + ".marker");
+        final TestingClusterEntrypointProcess clusterEntrypointProcess =
+                new TestingClusterEntrypointProcess(markerFile);
+        try {
+            clusterEntrypointProcess.startProcess();
+            final long pid = clusterEntrypointProcess.getProcessId();
+            assertTrue("Cannot determine process ID", pid != -1);
+
+            // wait for the marker file to appear, which means the process is up properly
+            TestJvmProcess.waitForMarkerFile(markerFile, 30000);
+
+            TestJvmProcess.killProcessWithSigTerm(pid);
+
+            CommonTestUtils.waitUntilCondition(
+                    () -> !markerFile.exists(),
+                    Deadline.fromNow(Duration.ofSeconds(3)),
+                    "markerFile should be deleted in closeAsync shutdownHook");
```

Comment:
We could simplify this slightly via

```suggestion
            clusterEntrypointProcess.waitFor();
            assertFalse(markerFile.exists());
```

Reply:
That way we don't use busy loop waiting.

Reply:
I am aware of `clusterEntrypointProcess.waitFor()`, but I am not pretty sure about whether it is reliable to depend on the `Process#waitFor`. So I am using the `CommonTestUtils.waitUntilCondition` before.

I have changed to `clusterEntrypointProcess.waitFor()` and run more than 100 times locally. It always passes. I will update this PR.